File: src/falcon_core/math/vector.py
"""Vector control parameters for feature optimizers."""

from typing import TYPE_CHECKING, cast

from ..dependancies import deepcopy, np, overload
from ..generic import Jsonable
from ..physics.device_structures.base_connection import BaseConnection
from ..physics.units import SymbolUnit, Units
from .point import Point
from .quantity import Quantity

if TYPE_CHECKING:
    from ..communications.voltage_states import DeviceVoltageStates
    from ..physics.device_structures.base_connection import BaseConnections
    from ..physics.units import SymbolUnit
    from .point import RawPoint, RawPointWUnits


class Vector(Jsonable):
    """A point in device gate space. And is N dimensional."""

    _end: "Point"
    _start: "Point"
    _connections: list[BaseConnection]
    _unit: "SymbolUnit"

    @overload
    def __init__(
        self,
        end_point: "RawPointWUnits",
    ): ...
    @overload
    def __init__(
        self,
        end_point: "RawPointWUnits",
        start_point: "RawPointWUnits",
    ): ...

    @overload
    def __init__(
        self,
        end_point: "RawPoint",
        start_point: None,
        unit: "SymbolUnit",
    ): ...
    @overload
    def __init__(
        self,
        end_point: "RawPoint",
        start_point: "RawPoint",
        unit: "SymbolUnit",
    ): ...

    @overload
    def __init__(
        self,
        end_point: "Point",
    ): ...

    @overload
    def __init__(
        self,
        end_point: "Point",
        start_point: "Point",
    ): ...

    def __init__(
        self,
        end_point: "RawPointWUnits | RawPoint | Point",
        start_point: "RawPointWUnits | RawPoint | Point | None" = None,
        unit: "SymbolUnit | None" = None,
    ) -> None:
        """Constructs a point.

        Args:
            value : the value of the point
            unit: optional type unit for the point
        """
        if unit is None:
            unit = cast("SymbolUnit", Units.VOLT)
        if not isinstance(end_point, Point):
            first = list(end_point.values())[0]
            if isinstance(first, float):
                end_point = cast("RawPoint", end_point)
                end_point = Point(
                    end_point,
                    unit=unit,
                )
            else:
                end_point = cast("RawPointWUnits", end_point)
                if unit is not None:
                    for quant in end_point.values():
                        quant.convert_to(unit)
                end_point = Point(end_point)
        if start_point is None:
            start_point = Point(
                {conn: 0.0 for conn in end_point.connections},
                unit=unit,
            )
        elif not isinstance(start_point, Point):
            first = list(start_point.values())[0]
            if isinstance(first, float):
                start_point = cast("RawPoint", start_point)
                start_point = Point(
                    start_point,
                    unit=unit,
                )
            else:
                start_point = cast("RawPointWUnits", start_point)
                if unit is not None:
                    for quant in start_point.values():
                        quant.convert_to(unit)
                start_point = Point(start_point)
        self._end = end_point
        self._start = start_point
        self._connections = list(
            set(end_point.connections).union(set(start_point.connections))
        )
        self._unit = end_point.unit

    @property
    def end(self) -> Point:
        """Returns the point at the end."""
        return self._end

    @property
    def start(self) -> Point:
        """Returns the point at the start."""
        return self._start

    @property
    def end_quantities(self) -> "RawPointWUnits":
        """Returns the map of quantities."""
        return self._end._coordinates

    @property
    def end_map(self) -> "RawPoint":
        """Returns the map of the point."""
        return {comp: value.value for comp, value in self.end_quantities.items()}

    @property
    def start_quantities(self) -> "RawPointWUnits":
        """Returns the map of quantities."""
        return self._start._coordinates

    @property
    def start_map(self) -> "RawPoint":
        """Returns the map of the point."""
        return {comp: value.value for comp, value in self.start_quantities.items()}

    @property
    def connections(self) -> list[BaseConnection]:
        """Returns the connections for a selected point."""
        return self._connections

    @property
    def principle_connection(self) -> BaseConnection:
        """Returns the principle_connections that are the greatest."""
        big_conn = self.connections[0]
        big_value = self.end_map[big_conn] - self.start_map[big_conn]
        for conn in self.connections:
            temp_value = self.end_map[conn] - self.start_map[conn]
            if temp_value > big_value:
                big_value = temp_value
                big_conn = conn
        return big_conn

    @property
    def unit(self) -> "SymbolUnit":
        """Returns the unit of the point."""
        return self._unit

    @property
    def magnitude(self) -> float:
        """The magnitude of the vector."""
        return np.sqrt(
            sum(
                [
                    (end - start) ** 2
                    for end, start in zip(
                        self.end_map.values(), self.start_map.values()
                    )
                ]
            )
        )

    @overload
    def __getitem__(self, connection: BaseConnection) -> tuple[float, float]: ...

    @overload
    def __getitem__(
        self, connection: "BaseConnections"
    ) -> list[tuple[float, float]]: ...

    def __getitem__(
        self,
        connection: "BaseConnections[BaseConnection] | BaseConnection",
    ) -> tuple[float, float] | list[tuple[float, float]]:
        """Returns the value of the component.

        Args:
            connection : the connection
        """
        if isinstance(connection, BaseConnection):
            if connection not in self.connections:
                msg = f"The connection {connection} is not inside the Vector. Available keys are {self.connections}"
                raise KeyError(msg)
            return (self.end_map[connection], self.start_map[connection])
        outs: list[tuple[float, float]] = []
        for conn in connection:
            outs.append(self[conn])
        return outs

    def __setitem__(
        self,
        connection: "BaseConnections[BaseConnection] | BaseConnection",
        value: tuple[float, float],
    ) -> None:
        """Sets the value of the component.

        Args:
            connection : the connection
            value : the value (start, end)
        """
        if isinstance(connection, BaseConnection):
            self._end[connection] = Quantity(value[1], self.unit)
            self._start[connection] = Quantity(value[0], self.unit)
        else:
            for conn in connection:
                self._end[conn] = Quantity(value[1], self.unit)
                self._start[conn] = Quantity(value[0], self.unit)

    def __neg__(self) -> "Vector":
        """The negation of a point is a point."""
        start = deepcopy(self.start_quantities)
        self.translate_to_origin()
        new = Vector(
            end_point={conn: -1 * quan for conn, quan in self.start_quantities.items()},
        )
        return new.translate(start)

    def update_start_from_states(self, state: "DeviceVoltageStates") -> "Vector":
        """Updates the vector to start from the given DeviceVoltageStates
        Args:
            state: the new device voltage state
        Returns:
            the displaced vector which starts at the given state.
        """
        new_origin = {}
        for dv_state in state.states:
            connection = dv_state.connection
            new_origin[connection] = Quantity(dv_state.value, unit=self.unit)
        return self.translate(new_origin)

    def translate(
        self,
        point: "RawPointWUnits | RawPoint | Point",
        unit: "SymbolUnit | None" = None,
    ) -> "Vector":
        """Displaces the origin of a vector by a point.

        Args:
            point: the displacement
            unit: the unit if the displacement is not Quantities

        Returns:
            the displaced vector that is translated
        """
        if unit is None:
            unit = self.unit
        if not isinstance(point, Point):
            first = list(point.values())[0]
            if isinstance(first, float):
                point = cast("RawPoint", point)
                point = Point(
                    point,
                    unit=unit,
                )
            else:
                point = cast("RawPointWUnits", point)
                if unit is not None:
                    for quant in point.values():
                        quant.convert_to(unit)
                point = Point(point)

        all_components = set(point.connections).union(self.connections)
        start = deepcopy(self.start)
        end = deepcopy(self.end)
        for component in all_components:
            if component not in point.connections:
                point[component] = Quantity(0.0, self.unit)
            if component not in start.connections:
                start[component] = Quantity(0.0, self.unit)
            if component not in end.connections:
                end[component] = Quantity(0.0, self.unit)

        for component in point.coordinates.values():
            component.convert_to(self.unit)
        for component, value in point.coordinates.items():
            start[component] += value.value
        for component, value in point.coordinates.items():
            end[component] += value.value
        return Vector(
            end_point=end,
            start_point=start,
        )

    def translate_to_origin(self) -> "Vector":
        """Translates a vector to the origin."""
        return self.translate(
            {conn: -1 * value for conn, value in self.start_quantities.items()}
        )

    def extend(self, extension: int | float) -> "Vector":
        """Extends a vector in place with its start point anchored."""
        origin = deepcopy(self.start_map)
        displacement = {component: -value for component, value in origin.items()}
        unit_vector = self.translate(point=displacement)
        scaled_vector = unit_vector * extension
        return scaled_vector.translate(
            point=origin,
            unit=self.unit,
        )

    def shrink(self, shrink: int | float) -> "Vector":
        """Shrinks a vector in place with its start point anchored."""
        origin = deepcopy(self.start_map)
        displacement = {component: -value for component, value in origin.items()}
        unit_vector = self.translate(point=displacement)
        scaled_vector = unit_vector / shrink
        return scaled_vector.translate(
            point=origin,
            unit=self.unit,
        )

    @property
    def unit_vector(self) -> "Vector":
        """Generates the unit vector for the direction of the vector."""
        unit_vector = self.translate(
            point={
                component: -1 * value
                for component, value in self.start_quantities.items()
            }
        )
        return unit_vector / self.magnitude

    def __add__(self, other: "Vector") -> "Vector":
        """The addition of two points is a vector."""
        return Vector(
            start_point=self.start_quantities,
            end_point=other.translate(point=self.end_quantities).end_quantities,
        )

    def __sub__(self, other: "Vector") -> "Vector":
        """The subtraction of two vector is a vector."""
        return self + (-other)

    def __mul__(self, other: float | int) -> "Vector":
        """The multiplication of a vector by a scalar is a vector."""
        return Vector(
            end_point={
                connection: value * other for connection, value in self.end_map.items()
            },
            start_point={
                connection: value * other
                for connection, value in self.start_map.items()
            },
            unit=self.unit,
        )

    def __rmul__(self, other: float | int) -> "Vector":
        """The reverse multiplication of a vector by a scalar is a vector."""
        return other * self

    def __truediv__(self, other: float | int) -> "Vector":
        """The division of a vector by a scalar is a vector."""
        # return (1 / float(other)) * self
        return Vector(
            end_point={
                connection: value / float(other)
                for connection, value in self.end_map.items()
            },
            start_point={
                connection: value / float(other)
                for connection, value in self.start_map.items()
            },
            unit=self.unit,
        )

    def normalize(self) -> "Vector":
        """Returns the normalized vector starting at the anchored starting point."""
        return self.shrink(self.magnitude)

    def project(self, other: "Vector") -> "Vector":
        """Projects ourself onto the other vector."""
        if other.unit != self.unit:
            for quant in other.end_quantities.values():
                quant.convert_to(self.unit)
            for quant in other.start_quantities.values():
                quant.convert_to(self.unit)
        origin_other = other.translate(
            {comp: -value for comp, value in other.start_map.items()}
        )
        origin = self.start_map
        origin_self = self.translate(
            {comp: -value for comp, value in self.start_map.items()}
        )
        shared_components = set(self.connections).intersection(set(other.connections))
        end = {
            comp: origin_self.end_map[comp] * origin_other.end_map[comp]
            for comp in shared_components
        }
        out = Vector(
            end_point=end,
            unit=self.unit,
            start_point=None,
        )
        return out.translate(origin, unit=self.unit)

File: src/math/Vector.cpp
#include "falcon_core/math/Vector.hpp"

#include <cmath>
#include <stdexcept>

#include "falcon_core/generic/Pair.hpp"
#include "falcon_core/math/Quantity.hpp"
#include "falcon_core/physics/device_structures/Connection.hpp"
#include "falcon_core/physics/device_structures/Connections.hpp"

namespace falcon_core::math {

Vector::Vector() = default;
Vector::Vector(const PointSP& start, const PointSP& end)
    : _unit(end ? end->unit() : nullptr),
      _connections(std::make_shared<physics::device_structures::Connections>()),
      generic::Map<physics::device_structures::Connection,
                   generic::Pair<Quantity, Quantity>>() {
  if (!start || !end) {
    throw std::invalid_argument(
        "Vector: The start and end points must not be null.");
  }
  auto end_connections   = end->connections();
  auto start_connections = start->connections();
  for (const physics::device_structures::ConnectionSP& connection :
       *end_connections) {
    _connections->push_back(connection);
  }
  for (const physics::device_structures::ConnectionSP& connection :
       *start_connections) {
    if (!_connections->contains(connection)) {
      _connections->push_back(connection);
    }
  }
  for (const physics::device_structures::ConnectionSP connectionSP :
       *_connections) {
    QuantitySP first, second;
    if (start->contains(connectionSP)) {
      first = start->at(connectionSP);
      first->convert_to(_unit);
    } else {
      first = std::make_shared<Quantity>(0.0, _unit);
    }
    if (end->contains(connectionSP)) {
      second = end->at(connectionSP);
      second->convert_to(_unit);
    } else {
      second = std::make_shared<Quantity>(0.0, _unit);
    }
    insert(connectionSP,
           std::make_shared<generic::Pair<Quantity, Quantity>>(first, second));
  }
}
Vector::Vector(const PointSP& end) : Vector(std::make_shared<Point>(), end) {}

Vector::Vector(
    const generic::MapSP<physics::device_structures::Connection, Quantity>&
        start,
    const generic::MapSP<physics::device_structures::Connection, Quantity>& end)
    : Vector(std::make_shared<Point>(start), std::make_shared<Point>(end)) {}

Vector::Vector(
    const generic::MapSP<physics::device_structures::Connection, Quantity>& end)
    : Vector(
          std::make_shared<
              generic::Map<physics::device_structures::Connection, Quantity>>(),
          end) {}

Vector::Vector(
    const generic::MapSP<physics::device_structures::Connection, double>& start,
    const generic::MapSP<physics::device_structures::Connection, double>& end,
    falcon_core::physics::units::SymbolUnitSP                             unit)
    : Vector(std::make_shared<Point>(start, unit),
             std::make_shared<Point>(end, unit)) {}

Vector::Vector(
    const generic::MapSP<physics::device_structures::Connection, double>& end,
    falcon_core::physics::units::SymbolUnitSP                             unit)
    : Vector(
          std::make_shared<
              generic::Map<physics::device_structures::Connection, double>>(),
          end,
          unit) {}
Vector::Vector(const generic::MapSP<physics::device_structures::Connection,
                                    generic::Pair<Quantity, Quantity>> map)
    : generic::Map<physics::device_structures::Connection,
                   generic::Pair<Quantity, Quantity>>(*map) {}

const PointSP Vector::endPoint() const {
  PointSP result = std::make_shared<Point>();
  for (const auto pair : items()) {
    result->insert(pair->first(), pair->second()->second());
  }
  return result;
}
const PointSP Vector::startPoint() const {
  PointSP result = std::make_shared<Point>();
  for (const auto pair : items()) {
    result->insert(pair->first(), pair->second()->first());
  }
  return result;
}
const generic::MapSP<physics::device_structures::Connection, Quantity>
Vector::end_quantities() const {
  generic::MapSP<physics::device_structures::Connection, Quantity> result =
      std::make_shared<Map<physics::device_structures::Connection, Quantity>>();
  for (const auto pair : items()) {
    result->insert(pair->first(), pair->second()->second());
  }
  return result;
}
const generic::MapSP<physics::device_structures::Connection, Quantity>
Vector::start_quantities() const {
  generic::MapSP<physics::device_structures::Connection, Quantity> result =
      std::make_shared<Map<physics::device_structures::Connection, Quantity>>();
  for (const auto pair : items()) {
    result->insert(pair->first(), pair->second()->first());
  }
  return result;
}
const generic::MapSP<physics::device_structures::Connection, double>
Vector::end_map() const {
  generic::MapSP<physics::device_structures::Connection, double> result =
      std::make_shared<Map<physics::device_structures::Connection, double>>();
  for (const auto pair : items()) {
    result->insert(pair->first(), pair->second()->second()->value());
  }
  return result;
}
const generic::MapSP<physics::device_structures::Connection, double>
Vector::start_map() const {
  generic::MapSP<physics::device_structures::Connection, double> result =
      std::make_shared<Map<physics::device_structures::Connection, double>>();
  for (const auto pair : items()) {
    result->insert(pair->first(), pair->second()->first()->value());
  }
  return result;
}
const physics::device_structures::ConnectionsSP& Vector::connections() const {
  return _connections;
}
const falcon_core::physics::units::SymbolUnitSP& Vector::unit() const {
  return _unit;
}
const physics::device_structures::ConnectionSP Vector::principle_connection()
    const {
  physics::device_structures::ConnectionSP big_conn = connections()->at(0);
  double big_value = (*at(big_conn)->second() - at(big_conn)->first())->value();
  for (const physics::device_structures::ConnectionSP& conn : *connections()) {
    double temp = (*at(conn)->second() - at(conn)->first())->value();
    if (temp > big_value) {
      big_value = temp;
      big_conn  = conn;
    }
  }
  return big_conn;
}
const double Vector::magnitude() const {
  double sum = 0.0;
  for (const physics::device_structures::ConnectionSP& conn : *_connections) {
    double diff = (*at(conn)->second() - at(conn)->first())->value();
    sum += diff * diff;
  }
  return std::sqrt(sum);
}

VectorSP Vector::operator+(const VectorSP& other) const {
  auto new_start = startPoint()->operator+(other->startPoint());
  auto new_end   = endPoint()->operator+(other->endPoint());
  return std::make_shared<Vector>(new_start, new_end);
}

VectorSP Vector::operator-(const VectorSP& other) const {
  auto new_start = startPoint()->operator-(other->startPoint());
  auto new_end   = endPoint()->operator-(other->endPoint());
  return std::make_shared<Vector>(new_start, new_end);
}

VectorSP Vector::operator*(double scalar) const {
  auto new_start = startPoint()->operator*(scalar);
  auto new_end   = endPoint()->operator*(scalar);
  return std::make_shared<Vector>(new_start, new_end);
}

VectorSP Vector::operator/(double scalar) const {
  auto new_start = startPoint()->operator/(scalar);
  auto new_end   = endPoint()->operator/(scalar);
  return std::make_shared<Vector>(new_start, new_end);
}

VectorSP Vector::operator-() const {
  auto new_start = startPoint()->operator-();
  auto new_end   = endPoint()->operator-();
  return std::make_shared<Vector>(new_start, new_end);
}
const VectorSP Vector::translate(const PointSP& point) const {
  if (!point) {
    throw std::invalid_argument(
        "Vector: The other point to translate by should not be null.");
  }
  PointSP startPoint = *this->startPoint() + point;
  PointSP endPoint   = *this->endPoint() + point;
  return std::make_shared<Vector>(startPoint, endPoint);
}
const VectorSP Vector::translate(
    const generic::MapSP<physics::device_structures::Connection, double> point,
    const physics::units::SymbolUnitSP unit) const {
  if (!point || !unit) {
    throw std::invalid_argument(
        "Vector: The other map and unit to translate by should not be null.");
  }
  PointSP adjustment = std::make_shared<Point>(point, unit);
  return translate(adjustment);
}
const VectorSP Vector::translate(
    const generic::MapSP<physics::device_structures::Connection, Quantity>
        point) const {
  if (!point) {
    throw std::invalid_argument(
        "Vector: The other map and unit to translate by should not be null.");
  }
  PointSP adjustment = std::make_shared<Point>(point);
  return translate(adjustment);
}
const VectorSP Vector::translate_to_origin() const {
  return translate(-*startPoint());
}
const VectorSP Vector::update_start_from_states(
    const communications::voltage_states::DeviceVoltageStatesSP& state) const {
  if (!state) {
    throw std::invalid_argument("Vector: The start state should not be null.");
  }
  VectorSP originVector = translate_to_origin();
  return originVector->translate(state->to_point());
}
const VectorSP Vector::unit_vector() const {
  VectorSP origin = translate_to_origin();
  return *origin / magnitude();
}
const VectorSP Vector::extend(const double& extension) const {
  VectorSP origin = translate_to_origin();
  VectorSP scaled = *origin * extension;
  return scaled->translate(startPoint());
}
const VectorSP Vector::extend(const int& extension) const {
  return extend(double(extension));
}
const VectorSP Vector::shrink(const double& shrink) const {
  VectorSP origin = translate_to_origin();
  VectorSP scaled = *origin / shrink;
  return scaled->translate(startPoint());
}
const VectorSP Vector::shrink(const int& shrink) const {
  return extend(double(shrink));
}
const VectorSP Vector::normalize() const { return shrink(magnitude()); }
void           Vector::update_unit(const physics::units::SymbolUnitSP& unit) {
  for (auto pair : items()) {
    pair->second()->first()->convert_to(unit);
    pair->second()->second()->convert_to(unit);
  }
}
const VectorSP Vector::project(const VectorSP& other) const {
  if (!other) {
    throw std::invalid_argument("Vector: The projection vector not be null.");
  }
  VectorSP clone = std::make_shared<Vector>(startPoint(), endPoint());
  clone->update_unit(other->unit());
  PointSP ourEnd   = clone->translate_to_origin()->endPoint();
  PointSP otherEnd = other->translate_to_origin()->endPoint();
  physics::device_structures::ConnectionsSP shared =
      connections()->intersection(other->connections());
  generic::MapSP<physics::device_structures::Connection, Quantity> rawPoint;
  for (const physics::device_structures::ConnectionSP& conn : *shared) {
    rawPoint->insert(conn, *ourEnd->at(conn) * otherEnd->at(conn));
  }
  VectorSP result = std::make_shared<Vector>(rawPoint);
  return result->translate(startPoint());
}
bool Vector::operator==(const Vector& other) const {
  return (*unit() == *other.unit()) &&
         (*connections() == *other.connections()) &&
         generic::Map<physics::device_structures::Connection,
                      generic::Pair<Quantity, Quantity>>::operator==(other);
}
bool Vector::operator!=(const Vector& other) const { return !(*this == other); }
}  // namespace falcon_core::math

CEREAL_REGISTER_TYPE(falcon_core::math::Vector)
using PQQ = falcon_core::generic::Pair<falcon_core::math::Quantity,
                                       falcon_core::math::Quantity>;
using MBD = falcon_core::generic::
    Map<falcon_core::physics::device_structures::Connection, PQQ>;
CEREAL_REGISTER_TYPE(PQQ)
CEREAL_REGISTER_TYPE(MBD)
CEREAL_REGISTER_POLYMORPHIC_RELATION(falcon_core::generic::Song, PQQ)
CEREAL_REGISTER_POLYMORPHIC_RELATION(falcon_core::generic::Song, MBD)
CEREAL_REGISTER_POLYMORPHIC_RELATION(MBD, falcon_core::math::Vector)
